### DISCUSSION TITLE
Developing with self-signed certificate?

### DISCUSSION THREAD

**Discussion by joelricci** [MatthiasValvekens/pyHanko#221] [Q&A] [answered]
Hi, is there any way to test and develop with pyHanko using a self-signed certificate? I realise that the generated PDF won't have a valid signature, but for testing and learning this would be ok.

I followed the example here: https://github.com/MatthiasValvekens/pyHanko/blob/master/docs/lib-guide/signing.rst#creating-pades-signatures. The problem I have is that I want to experiment with PAdES signatures but the certificate can't be validated:

`pyhanko.sign.general.SigningError: The signer's certificate could not be validated`

I couldn't find any information on how to do this.

Thanks

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi @joelricci,

Excellent question. There are several reasons why testing with a _standalone_ self-signed certificate doesn't really work all that well with PAdES and other advanced signature standards:

 - In real life, the signer's certificate alone doesn't mean all that much, it needs to chain up to a root that the validator trusts. This may require passing through multiple intermediate CAs, and embedding those certs  into the document is considered good practice. Having this set-up in your testing/dev environment is convenient.
 - However, the biggest reason is revocation info embedding: one of the ingredients in extending the lifetime of a signature beyond that of its certificate involves embedding validation data together with the signature. These are things like OCSP responses, CRLs, etc. In a typical scenario, embedding this information for the signer's certificate alone is actually not enough. The intermediate CAs are subject to such checks as well. If you add timestamping authorities and CRL issuers into the mix, it gets even more complicated. The easiest way to ensure that all relevant info is actually collected properly (assuming you know more or less what the validator expects) is to perform that validation at signing time and embed all the data collected during that process. There's no way to replicate this with a self-signed cert only.

That said, obviously I'm not advocating that you use your production keys for testing and development! Instead, you may want to look into a tool like [Certomancer](https://github.com/MatthiasValvekens/certomancer), which was designed for exactly this use case: it can generate test certs, set up OCSP responders and CRLs, and function as a TSA (both as actual network services and in your testing process). Basically, it allows you to set up a PKI environment that closely mirrors the user-facing architecture used by "real" CAs. All you have to do is supply the keys and some config, and configure your `ValidationContext` at testing time to accept the "fake" root(s) underpinning your Certomancer PKI.

There are more heavy-handed solutions that involve deploying actual CA software in your testing environment, but chances are that you don't need that.

FWIW, [pyHanko's test suite](https://github.com/MatthiasValvekens/pyHanko/tree/master/pyhanko_tests) makes very heavy use of Certomancer.

**Reply by joelricci** [reply]
Excellent answer. I will try to grasp what all this means and also have a look at Certomancer and the test suit. I might come back with more questions :)

Kind regards
Joel
